### PR TITLE
feat: support trace with executables

### LINKED ISSUE #63: Trace support for executables
One use case: https://github.com/notaryproject/notation/pull/654#discussion_r1180080763

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Resolves #63

### CODE REVIEW

**Comment by Wwwsylvia** [comment]
A general issue: 
Most users don't use native store directly. Instead, they use [`DynamicStore`](https://pkg.go.dev/github.com/oras-project/oras-credentials-go#DynamicStore) which may internally leverage native store. We should make sure that the calls to native store through `DynamicStore` can also be traced.

File: internal/executer/executer.go [hunk 1]
```diff
@@ -49,7 +51,14 @@ func New(name string) Executer {
 
 // Execute operates on an executable binary and supports context.
 func (c *executable) Execute(ctx context.Context, input io.Reader, action string) ([]byte, error) {
+	trace := trace.GetTraceHooksFromContext(ctx)
+	if trace != nil {
+		trace.ExecuteStart(c.name, action)
+	}
 	cmd := exec.CommandContext(ctx, c.name, action)
+	if trace != nil {
+		trace.ExecuteDone(c.name, action)
```
Comment (shizhMSFT) [inline internal/executer/executer.go:60]: What if `trace.ExecuteDone` is nil?

Comment (wangxiaoxuan273) [reply internal/executer/executer.go:60]: Added the check. Thanks for catching it.

File: internal/executer/executer.go [hunk 2]
```diff
@@ -49,7 +51,14 @@ func New(name string) Executer {
...
+	}
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply internal/executer/executer.go:61]: Moved the trace hook executions around `cmd.Output()`

File: trace/trace.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+type ExecutableTrace struct {
+	// ExecuteStart is called before the execution of the executable. The
+	// executableName parameter is the name of the credential helper executable
+	// used with NativeStore. The action parameter is one of "store", "get" and
+	// "erase".
+	//
+	// Reference:
+	//   - https://docs.docker.com/engine/reference/commandline/login#credentials-store
+	ExecuteStart func(executableName string, action string)
+	// ExecuteEnd is called after the execution of an executable completes.
```
Comment (Two-Hearts) [inline trace/trace.go:37]: nit:
need an extra line here and rename
```suggestion

	// ExecuteDone is called after the execution of an executable completes.
```

File: trace/trace.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+package trace
+
+import (
+	"context"
+)
+
+// executableTraceHookKey is a value key used to retrieve the ExecutableTrace
+// from Context.
+type executableTraceContextKey struct{}
```
Comment (Two-Hearts) [inline trace/trace.go:24]: how about:
```suggestion
type contextKey int

type contextKeyExecutableTrace contextKey
```

File: trace/trace.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	//
+	// Reference:
+	//   - https://docs.docker.com/engine/reference/commandline/login#credentials-store
+	ExecuteDone func(executableName string, action string, err error)
+}
+
+// GetTraceHooksFromContext returns the ExecutableTrace associated with the context.
+// If none, it returns nil.
+func ContextExecutableTrace(ctx context.Context) *ExecutableTrace {
+	trace, _ := ctx.Value(executableTraceContextKey{}).(*ExecutableTrace)
```
Comment (Two-Hearts) [inline trace/trace.go:50]: If taking the suggested change above, then this line becomes:
```suggestion
	trace, _ := ctx.Value(contextKeyExecutableTrace).(*ExecutableTrace)
```

File: trace/trace.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	return trace
+}
+
+// WithTraceHooks takes a Context and an ExecutableTrace, and returns a Context with
+// the ExecutableTrace added as a Value.
+func WithExecutableTrace(ctx context.Context, trace *ExecutableTrace) context.Context {
+	if trace != nil {
+		ctx = context.WithValue(ctx, executableTraceContextKey{}, trace)
```
Comment (Two-Hearts) [inline trace/trace.go:58]: And this line becomes:
```suggestion
		ctx = context.WithValue(ctx, contextKeyExecutableTrace, trace)
```

File: trace/trace.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+// executableTraceHookKey is a value key used to retrieve the ExecutableTrace
+// from Context.
+type executableTraceContextKey struct{}
+
+// ExecutableTrace is a set of hooks used to trace the execution of the binary
+// executable used with NativeStore. Any particular hook may be nil.
+type ExecutableTrace struct {
+	// ExecuteStart is called before the execution of the executable. The
+	// executableName parameter is the name of the credential helper executable
+	// used with NativeStore. The action parameter is one of "store", "get" and
```
Comment (Two-Hearts) [inline trace/trace.go:31]: Just curious, what would happen if I passed a string other than [store, get, erase]? 
(Personally, I'd prefer to make variable like `action` to enum instead of a string as it only takes 3 valid values.)

File: trace/trace.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply trace/trace.go:24]: Resolved this comment and the related ones by offline discussion.

File: trace/trace.go (hunk 7 above)
Comment (shizhMSFT) [reply trace/trace.go:31]: `ExecuteStart` is for logging purpose, and is always invoked by `oras`. If a user hook receives action other than [store, get, erase]. It's probably a programming bug.

File: trace/trace.go [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+	return trace
+}
+
+// WithExecutableTrace takes a Context and an ExecutableTrace, and returns a Context with
+// the ExecutableTrace added as a Value.
+func WithExecutableTrace(ctx context.Context, trace *ExecutableTrace) context.Context {
+	if trace != nil {
+		ctx = context.WithValue(ctx, executableTraceContextKey{}, trace)
+	}
+	return ctx
```
Comment (shizhMSFT) [inline trace/trace.go:61]: nit: code style:
```suggestion
	if trace == nil {
		return ctx 
	}
	return context.WithValue(ctx, executableTraceContextKey{}, trace)
```

File: trace/trace.go (hunk 7 above)
Comment (wangxiaoxuan273) [reply trace/trace.go:31]: For your information, if you pass a string other than [store, get, erase], the behavior is up to the executable. For docker credential helper executables, an error will be thrown. @Two-Hearts

File: internal/executer/executer.go [hunk 9]
```diff
@@ -49,10 +51,17 @@ func New(name string) Executer {
 
 // Execute operates on an executable binary and supports context.
 func (c *executable) Execute(ctx context.Context, input io.Reader, action string) ([]byte, error) {
+	trace := trace.ContextExecutableTrace(ctx)
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply internal/executer/executer.go:54]: Moved it to around the usage of trace.

File: trace/trace.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+import (
+	"context"
+)
+
+// executableTraceContextKey is a value key used to retrieve the ExecutableTrace
+// from Context.
+type executableTraceContextKey struct{}
+
+// ExecutableTrace is a set of hooks used to trace the execution of the binary
+// executable used with NativeStore. Any particular hook may be nil.
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply trace/trace.go:27]: I chose to delete 'with NativeStore'.

File: trace/trace.go (hunk 8 above)
Comment (wangxiaoxuan273) [reply trace/trace.go:61]: changed accordingly.

File: trace/trace.go [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+
+// compose takes an oldTrace and modifies the existing trace to include
+// the hooks defined in the oldTrace. The hooks in the existing trace will
+// be called first.
+func (trace *ExecutableTrace) compose(oldTrace *ExecutableTrace) {
+	oldStart := oldTrace.ExecuteStart
+	if oldStart != nil {
+		start := trace.ExecuteStart
+		trace.ExecuteStart = func(executableName, action string) {
+			start(executableName, action)
```
Comment (shizhMSFT) [inline trace/trace.go:78]: What if `start` is `nil`?

File: trace/trace.go [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+			oldStart(executableName, action)
+		}
+	}
+	oldDone := oldTrace.ExecuteDone
+	if oldDone != nil {
+		done := trace.ExecuteDone
+		trace.ExecuteDone = func(executableName, action string, err error) {
+			done(executableName, action, err)
```
Comment (shizhMSFT) [inline trace/trace.go:86]: What if `done` is `nil`?

File: trace/trace.go [hunk 13]
```diff
@@ -0,0 +1,90 @@
...
+
+// WithExecutableTrace takes a Context and an ExecutableTrace, and returns a Context with
+// the ExecutableTrace added as a Value. If the Context has a previously added trace,
+// the hooks defined in the new trace will be added in addition to the previous ones.
+// The recent hooks will be called first.
+func WithExecutableTrace(ctx context.Context, trace *ExecutableTrace) context.Context {
+	if trace == nil {
+		return ctx
+	}
+	oldTrace, _ := ctx.Value(executableTraceContextKey{}).(*ExecutableTrace)
```
Comment (shizhMSFT) [inline trace/trace.go:63]: `ContextExecutableTrace()` should be reused for this line.

**Comment by TerryHowe** [comment]
I just wonder why this isn't being added to oras-go

**Comment by wangxiaoxuan273** [comment]
> I just wonder why this isn't being added to oras-go

@TerryHowe  It will be added to `oras-go` soon.

File: trace/trace.go (hunk 13 above)
Comment (wangxiaoxuan273) [reply trace/trace.go]: Resued `ContextExecutableTrace()`

File: trace/trace.go (hunk 12 above)
Comment (wangxiaoxuan273) [reply trace/trace.go]: Added the check and refactored this part of code.

File: trace/trace.go (hunk 11 above)
Comment (wangxiaoxuan273) [reply trace/trace.go]: Added the check and refactored this part of code.

**Review by shizhMSFT — APPROVED** [review]
LGTM with suggestions

File: native_store_test.go [hunk 14]
```diff
@@ -16,22 +16,26 @@ limitations under the License.
...
+	"bytes"
 	"context"
 	"encoding/json"
 	"fmt"
 	"io"
 	"strings"
 	"testing"
 
+	"github.com/oras-project/oras-credentials-go/trace"
+
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply native_store_test.go]: removed.

File: trace/trace.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+// If none, it returns nil.
+func ContextExecutableTrace(ctx context.Context) *ExecutableTrace {
+	trace, _ := ctx.Value(executableTraceContextKey{}).(*ExecutableTrace)
+	return trace
+}
+
+// WithExecutableTrace takes a Context and an ExecutableTrace, and returns a Context with
+// the ExecutableTrace added as a Value. If the Context has a previously added trace,
+// the hooks defined in the new trace will be added in addition to the previous ones.
+// The recent hooks will be called first.
```
Comment (Wwwsylvia) [inline trace/trace.go:58]: nit: Comment line length exceeds 80.

File: native_store_test.go [hunk 16]
```diff
@@ -185,3 +209,133 @@ func TestNewDefaultNativeStore(t *testing.T) {
...
+		t.Fatalf("trace test ns.Delete fails: %v", err)
+	}
+	bufferContent = buffer.String()
+	if bufferContent != "test trace, start the execution of executable testExecuter with action erase test trace, completed the execution of executable testExecuter with action erase and got err <nil>" {
+		t.Fatalf("incorrect buffer content: %s", bufferContent)
+	}
+}
+
+// This test ensures that a nil trace will not cause an error.
+func TestNativeStore_noTrace(t *testing.T) {
```
Comment (Wwwsylvia) [inline native_store_test.go:260]: This test case looks identical with `TestNativeStore_basicAuth`, which tests the scenario where the context carries no trace. You may also want to test the case where the context carries an empty trace.

Comment (wangxiaoxuan273) [reply native_store_test.go:260]: Added the empty trace test case. Good suggestion, thanks.

File: trace/trace.go (hunk 15 above)
Comment (wangxiaoxuan273) [reply trace/trace.go:58]: shrinked the line length

**Review by Wwwsylvia — APPROVED** [review]
LGTM
